src/pybind/wrapper.cpp

```diff
@@ -0,0 +1,221 @@
+        // x = x + f(x) * dt
+        py::exec(R"(
+                from nmodl.ode import forwards_euler2c
+                exception_message = ""
+                try:
+                    solution = forwards_euler2c(equation_string, dt_var, vars, function_calls)
+                except Exception as e:
+                    # if we fail, fail silently and return empty string
+                    solution = ""
+                    exception_message = str(e)
+            )",
+                 py::globals(),
+                 locals);
+    } else if (method == codegen::naming::CNEXP_METHOD) {
+        // replace x' = f(x) differential equation
+        // with analytic solution for x(t+dt) in terms of x(t)
+        // x = ...
+        py::exec(R"(
+                from nmodl.ode import integrate2c
+                exception_message = ""
+                try:
+                    solution = integrate2c(equation_string, dt_var, vars,
+                                           use_pade_approx)
+                except Exception as e:
+                    # if we fail, fail silently and return empty string
+                    solution = ""
+                    exception_message = str(e)
+            )",
+                 py::globals(),
+                 locals);
+    } else {
+        // nothing to do, but the caller should know.
+        return;
+    }
+    solution = locals["solution"].cast<std::string>();
+    exception_message = locals["exception_message"].cast<std::string>();
+}
+
+void AnalyticDiffExecutor::operator()() {
+    auto locals = py::dict("expressions"_a = expressions, "vars"_a = used_names_in_block);
+    py::exec(R"(
+                            from nmodl.ode import differentiate2c
+                            exception_message = ""
+                            try:
+                                rhs = expressions[-1].split("=", 1)[1]
+                                solution = differentiate2c(rhs,
+                                                           "v",
+                                                           vars,
+                                                           expressions[:-1]
+                                           )
+                            except Exception as e:
+                                # if we fail, fail silently and return empty string
+                                solution = ""
+                                exception_message = str(e)
+                        )",
+             py::globals(),
+             locals);
+    solution = locals["solution"].cast<std::string>();
+    exception_message = locals["exception_message"].cast<std::string>();
+}
+
+SolveLinearSystemExecutor* create_sls_executor_func() {
+    return new SolveLinearSystemExecutor();
+}
+
+SolveNonLinearSystemExecutor* create_nsls_executor_func() {
+    return new SolveNonLinearSystemExecutor();
+}
+
+DiffeqSolverExecutor* create_des_executor_func() {
+    return new DiffeqSolverExecutor();
+}
```

Comment:
most likely I don't have global workflow of the code but curious why it was not possible return smart ptr to avoid explicit call to `destroy*`.

Reply:
passing smart pointers through dynamically loaded libraries.. hmm.. not sure what would happen! The reason why I insist on the destructor being called from within the `.so` is because I want to make sure destructor code is definitely reachable. I don't claim I fully understand the details. But here is some relevant documentation:
http://www.faqs.org/docs/Linux-mini/C++-dlopen.html#loadingclasses
(not end of sec 3.3)

I would have liked to have all implementations of `PythonExecutor` to have the same interface rather than these structs where one must set the fields from outside, but what I cam up with was ugly.. so I deferred this for later in the interest of time.

Reply:
yup, that's fine!